subsystem/apiclient/kubeapi.go

```diff
@@ -0,0 +1,288 @@
+			HTTPProxy:                swag.StringValue(cParams.HTTPProxy),
+			HTTPSProxy:               swag.StringValue(cParams.HTTPSProxy),
+			NoProxy:                  swag.StringValue(cParams.NoProxy),
+			UserManagedNetworking:    swag.BoolValue(cParams.UserManagedNetworking),
+			AdditionalNtpSource:      swag.StringValue(cParams.AdditionalNtpSource),
+			PullSecretRef:            pullSecretRef,
+		},
+	}
+	if deployErr := kc.client.Create(ctx, &c); deployErr != nil {
+		return nil, errors.Wrapf(deployErr, "failed to deploy cluster")
+	}
+
+	key := types.NamespacedName{
+		Name:      metaName,
+		Namespace: kc.namespace,
+	}
+	return kc.getClusterWithRetries(ctx, key)
+}
+
+func (kc *KubeAPIClient) deployPullSecret(
+	ctx context.Context,
+	name, pullSecret string,
+) (*corev1.SecretReference, error) {
+
+	secretName := fmt.Sprintf("%s-secret", name)
+	secretRef := &corev1.SecretReference{
+		Name:      secretName,
+		Namespace: kc.namespace,
+	}
+
+	key := types.NamespacedName{
+		Name:      secretName,
+		Namespace: kc.namespace,
+	}
+	secret := &corev1.Secret{}
+	if getSecretErr := kc.client.Get(ctx, key, secret); runtimeclient.IgnoreNotFound(getSecretErr) != nil {
+		return nil, getSecretErr
+	} else if getSecretErr == nil {
+		return secretRef, nil
+	}
+
+	if pullSecret == "" {
+		return nil, nil
+	}
+
+	data := map[string]string{"pullSecret": pullSecret}
+	secretData := make(map[string][]byte)
+	for k, v := range data {
+		secretData[k] = []byte(base64.StdEncoding.EncodeToString([]byte(v)))
+	}
+
+	secret = &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: APIVersion,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: kc.namespace,
+		},
+		Data: secretData,
+	}
+	if err := kc.client.Create(ctx, secret); err != nil {
+		return nil, err
+	}
+
+	return secretRef, nil
+}
+
+func (kc *KubeAPIClient) verifyClusterStatus(ctx context.Context, key types.NamespacedName) error {
+	cluster := &adiiov1alpha1.Cluster{}
+	if err := kc.client.Get(ctx, key, cluster); err != nil {
+		return err
+	}
+	if cluster.Status.Error != "" {
+		return errors.Errorf("cluster status error: %s", cluster.Status.Error)
+	}
+	return nil
+}
+
+func (kc *KubeAPIClient) getClusterWithRetries(
+	ctx context.Context,
+	key types.NamespacedName,
+) (*models.Cluster, error) {
+
+	var cluster *models.Cluster
+	var getClusterErr error
+	for i := 0; i < 10; i++ {
+		if verifyStatusErr := kc.verifyClusterStatus(ctx, key); verifyStatusErr != nil {
+			return nil, verifyStatusErr
+		}
+		if cluster, getClusterErr = kc.getClusterByKey(key); getClusterErr == gorm.ErrRecordNotFound {
+			time.Sleep(time.Millisecond * 500)
+			continue
+		}
+		break
+	}
+	if getClusterErr != nil {
+		return nil, errors.Wrapf(getClusterErr, "failed to get cluster")
+	}
+	return cluster, nil
+}
+
+func (kc *KubeAPIClient) getClusterByKey(key types.NamespacedName) (*models.Cluster, error) {
```

Comment:
I'm not sure if we should take the cluster directly from the DB or get it with KubeAPI and translate to `model.Cluster` this way we will hide translation bugs. but it can probably be made in a different PR.